### ISSUE TITLE
Proper non-trigger-based support for generic/custom events

### ISSUE THREAD

**Issue by jlfarris91** [wurstscript/WurstScript#895]
I come from a heavy .Net background and really wanted to use lightweight events in Wurst that don't rely on triggers. I've been using this in my code base and thought it might be nice to have proper Wurst support for something similar. I'm still new with Wurst so there may be a better reason to use triggers that I'm unaware of.

Any thoughts or suggestions for me?

```
// ============================================================================
public interface GenericArgsEventHandler<TSender, TArgs>
  function callback(TSender sender, TArgs args)

// ============================================================================
public class GenericArgsEvent<TSender, TArgs>
  private LinkedList<GenericArgsEventHandler<TSender, TArgs>> _handlers

  // --------------------------------------------------------------------------
  ondestroy
    if (_handlers != null)
      for handler in _handlers
        destroy handler
      destroy _handlers
      _handlers = null

  // --------------------------------------------------------------------------
  function addListener(GenericArgsEventHandler<TSender, TArgs> handler) returns GenericArgsEventHandler<TSender, TArgs>
    if (handler == (0 castTo GenericArgsEventHandler<TSender, TArgs>))
      return handler
    if (_handlers == null)
      _handlers = new LinkedList<GenericArgsEventHandler<TSender, TArgs>>()
    _handlers.add(handler)
    return handler

  // --------------------------------------------------------------------------
  function removeListener(GenericArgsEventHandler<TSender, TArgs> handler) returns GenericArgsEventHandler<TSender, TArgs>
    if (handler != (0 castTo GenericArgsEventHandler<TSender, TArgs>) and _handlers != null)
      _handlers.remove(handler)
    return 0 castTo GenericArgsEventHandler<TSender, TArgs>

  // --------------------------------------------------------------------------
  function invoke(TSender sender, TArgs args)
    if (_handlers != null)
      for handler in _handlers
        handler.callback(sender, args)
```

Below is a contrived example that relies on knowing the sender of the event and arguments passed to the event. In this example we have a Killer object that is metadata for some unit that can raise an event for when the unit gets multiple kills and a MultiKillReporter object that is responsible for  listening to the gotMultikill event on every Killer object and reporting a message to all other players when the Killer gets a multikill.

```

class Killer
  GenericArgsEvent<Killer, int> gotMultikill

  construct()
    gotMultikill = new GenericArgsEvent<Killer, int>()

  ondestroy
    destroy gotMultikill

  function killUnits(varargs unit unitsToKill)
    var kills = 0
    for unitToKill in unitsToKill
      killTheUnit(unitToKill)
      kills++
    if (kills > 1)    
      gotMultikill.invoke(this, kills)

class MultikillReporter
  private LinkedList<Killer> _killers

  // This would probably be cleaner with a HashMap instead
  private LinkedList<GenericArgsEventHandler<Killer, int>> _handlers

  construct()
    _killers = new LinkedList<Killer>()
    _handlers = new LinkedList<GenericArgsEventHandler<Killer, int>>()

  ondestroy
    for killer in _killers
      unregisterKiller(killer)
    destroy _killers
    destroy _handlers

  function registerKiller(Killer killer)
    _killers.add(killer)
    var handler = _killer.gotMultikill.addListener() (Killer killer, int kills) ->
      reportMultikill(killer, kills)
    _handlers.add(handler)

  function unregisterKiller(Killer killer)
    var i = 0
    for k in _killers
      if (k == killer)
        _killers.remove(k)
        _killer.gotMultikill.removeListener(_handlers.removeAt(i))
        return
      i++

  private function reportMultikill(Killer killer, int kills)
    if (kills == 2)
      report(killer, "DOUBLE KILL")
    else if (kills == 3)
      report(killer, "TRIPLE KILL")
    else if (kills == 4)
      report(killer, "QUADRA KILL")
    else
      report(killer, "GOD LIKE")

  private function report(Killer killer, string message)
    // Do something here to report the multikill to all players

```

**Comment by Frotty** [comment]
In wurst you use closures.
See https://wurstlang.org/tutorials/getclose.html

**Comment by jlfarris91** [comment]
I'm using closures here. I just thought an object for custom events like this might be a nice addition to the standard library as an alternative to using triggers. I figured you guys would probably be able to provide something similar but more efficient than what I've written.

Thanks for the advice.

**Comment by Frotty** [comment]
Okay sorry, then I misunderstood your question and this should probably be an issue on the standard library instead of compiler.
It's not immediately obvious how your code should be used, please add an example.
I guess this is supposed to help newcomers if it were to replace triggers, but I don't see how right now and `int sender` looks suspicious.

In many cases though I think it is advisable to use a type specific closure.
For cases where you just want to pass functions around, https://github.com/theQuazz/wurst-lodash provides some nice generic closures.

**Comment by jlfarris91** [comment]
Thanks for your feedback, I made some changes. Also, thanks for the link; I'll check out that library when I get some time.

1. I removed the @inlines as per your suggestion
2. I changed the GenericArgsEventHandler<TArgs> to take 2 types, one for the sender and one for the arguments, removing the need for  `int sender` as per your suggestion
3. I replaced my previous code that included multiple classes with just the most important one, hopefully this will help clarify what I'm trying to achieve. My library still has 2 other event classes, one with a sender and no arguments (GenericEvent) and one with no sender and no arguments (AnonymousEvent). I'm open to suggestions for better names for these classes XD
4. I added an example. I can share my actual project with you if you are interested in seeing how I'm actually using these classes.
